includes/utilities.php

```diff
@@ -150,14 +176,27 @@ function insert_redirect( $from, $to, $status_code, $post_id = 0 ) {
 	 */
 	$to = apply_filters( 'hm_redirects_pre_save_to_url', $to, $from, $status_code, $post_id );
 
+	// Parse out query string parameters and add them to the preserved list.
+	wp_parse_str( wp_parse_url( $from, PHP_URL_QUERY ), $query );
+	if ( is_array( $query ) && ! empty( $query ) ) {
+		add_preserved_query_string_parameters( array_keys( $query ) );
```

Comment:
I'm not following the logic here, when any redirect is inserted we globally store it's URL params with all other redirect params?

Reply:
Ok, I guess this is part of the "is the url param stored with the redirect-from or not", but as this is being added to a global list, I don't really see how that will work.

Reply:
Yeah it's not perfect by any means but I was trying to find a way to keep the existing functionality while still being able to preserve query string parameters during a redirect.

Current you could have redirects like the following:

- From `/?utm_campaign=fb-inbound` -> `/facebook-landing-page/`
- From `/old-source-url/posts?p=123` -> `/2018/02/11/hello-world/`

If we don't have a way to know what query string parameters are important for redirects then we could only ever match a redirect in the database by path. For speed this plugin uses `get_page_by_path` which does a direct (cached) SQL query to retrieve a single post rather than using `WP_Query`.

Given all of that this seemed to me like the only way to know which query string params should be removed before looking for matching redirects.

Just realised though that I would need to store key & value pairs for query string params used in redirects though.

So, we have a few options:

- balancing performance with backwards compatibility
- breaking back compat and not supporting query strings in URLs to redirect _from_
- potentially losing some performance to match multiple redirects by path using `WP_Query` and then checking some other data point for matching query string params